packages/client/examples/node.js/mutability.js

```diff
@@ -0,0 +1,51 @@
+import process from 'process'
+import minimist from 'minimist'
+import { Web3Storage, Name } from 'web3.storage'
+
+async function main () {
+  const args = minimist(process.argv.slice(2))
+  const token = args.token
+
+  if (!token) {
+    console.error('A token is needed. You can create one on https://web3.storage')
+    return
+  }
+
+  const client = new Web3Storage({ token, endpoint: args.endpoint })
+
+  // Create a private key, and key ID (the "name" used to resolve the current value)
+  const { id, privateKey } = await Name.keypair()
+
+  console.log(`\n🆔 Created key ID:\n${id}\n`)
+
+  // The value to publish
+  const value = '/ipfs/bafkreiem4twkqzsq2aj4shbycd4yvoj2cx72vezicletlhi7dijjciqpui'
+
+  // Create a new name record for the given value
+  const { record } = await Name.create(privateKey, null, value)
+  console.log(`💿 Created new IPNS record /ipns/${id} => ${value}:\n${record}\n`)
+
+  console.log('⏳ Publishing to Web3.Storage...')
+  await Name.publish(client, id, record)
+  console.log('✅ Done\n')
+
+  console.log('⏳ Resolving current value...')
+  const { value: curValue, record: curRecord } = await Name.resolve(client, id)
```

Comment:
I like to have Name exported as an util. But, I would expect `resolve` and `publish` to be part of the main client API. What are your thoughts here? I am with mixed feelings about this.

1. (+) It is good that it is isolated because people will need to use `Name.keypair()` and `Name.create()` to publish.
2. (+) We can probably isolate in the context of a shared dotstorage lib
2. (-) Passing the client into `Name.resolve()` and `Name.publish()` is strange IMHO
3. (-) Folks can resolve by receiving an ID from a third party, having to import Name for it

Reply:
I'd be happy to add `publish` and `resolve` to the main client as sugar for calling `Name.publish` and `Name.resolve`.

Reply:
I think it would really help to refine the terms. Here is what would make most sense to me:


```ts
// Creates new writable name.
declare function create():Promise<WritableName>
// Creates writable name from keypair.
declare function from(keypair:KeyPair):Promise<WritableName>
// Parses string name to readable name
declare function parse(name:string):Name

export interface Name {
  // binary representation of the identifier (hash digest of the public key)
  bytes: Uint8Array
  // returns string representation of the name
  toString(base?:BaseEncoder):string 
}

// Writable name is just a name with a signing key attached so it could be
// used for publishing
export interface WritableName extends Name {
  key: SigningKey
}

export interface SigningKey {
  sign(input:Uint8Array):Promise<Uint8Array>
}

// Creates initial revision for the given name name. Note this does not require // signing key because it is ok to create revision without it, publishing it would
// require key however.
declare function v0(name:Name, value:string):Revision

// Creates an incremented revision
declare function increment(revision:Revision, value:string):Revision

interface Revision {
  name: Name
  sequence: number
  value: string
}

// Publishes given revision to a service. Requires signing key
declare function publish(service:Service, revision, key:SigningKey}):Promise<Revision>

// Resolves last published revision from the service
declare function resolve(service:Service, name:Name):Promise<Revision>
```

Reply:
I've implemented this as designed above. I think it's a great idea to separate the name from the record. I've also separated this from the main bundle, so you use it like `import * as Name from 'web3.storage/name'`. This allows us to publish it as is and try it out without effecting the main library. Once we're happy with it and we've removed libp2p-crypto to minimise the bundle size we can add sugar to the main library.